internal/spec/spec.go

```diff
@@ -202,22 +203,23 @@ func (r RunnerSpec) ComposeUserData() (string, error) {
 			return "", fmt.Errorf("failed to generate userdata: %w", err)
 		}
 
-		lines := strings.Split(string(udata), "\n")
-		if len(lines) > 0 && strings.HasPrefix(lines[0], "#!") {
-			additionalCommands := []string{
-				// Create user 'runner' if it doesn't exist; '|| true' to ignore if user already exists
-				"sudo useradd -m " + defaults.DefaultUser + " || true",
-				// Create the runner home directory if it doesn't exist
-				"sudo mkdir -p /home/" + defaults.DefaultUser,
-				// Add user to sudoers
-				"sudo usermod -aG sudo " + defaults.DefaultUser,
-				// Check curl and tar are installed
-				"sudo apt-get update && sudo apt-get install -y curl tar",
-			}
-			lines = append(lines[:1], append(additionalCommands, lines[1:]...)...)
+		asBase64 := base64.StdEncoding.EncodeToString(udata)
+		scriptCommands := []string{
+			"sudo useradd -m " + defaults.DefaultUser + " || true",
+			// Create the runner home directory if it doesn't exist
+			"sudo mkdir -p /home/" + defaults.DefaultUser,
+			// Add user to sudoers
+			"sudo usermod -aG sudo " + defaults.DefaultUser,
+			// Check curl and tar are installed
+			"sudo apt-get update && sudo apt-get install -y curl tar",
+			// Install the runner
+			"echo " + asBase64 + " | base64 -d > /install_runner.sh",
+			"chmod +x /install_runner.sh",
+			"echo '" + defaults.DefaultUser + " ALL=(ALL) NOPASSWD:ALL' | sudo EDITOR='tee -a' visudo",
```

Comment:
no need for `visudo` here. Just write the file with the proper permissions. The `visudo` command is just meant to validate the `sudoers` file. If you're certain that the string is valid, you can just `echo`.

Reply:
also, most systems have a `/etc/sudoers.d/` folder. So you can just echo:

```bash
echo 'runner  ALL=(ALL) NOPASSWD:ALL' > /etc/sudoers.d/garm
```

Reply:
Updated! Thanks!